x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/ChunkedTrainedModelPersister.java

```diff
@@ -0,0 +1,208 @@
+        this.failureHandler = failureHandler;
+        this.extractedFields = extractedFields;
+    }
+
+    public void createAndIndexInferenceModelDoc(TrainedModelDefinitionChunk trainedModelDefinitionChunk) {
+        if (Strings.isNullOrEmpty(this.currentModelId.get())) {
+            failureHandler.accept(ExceptionsHelper.serverError(
+                "chunked inference model definition is attempting to be stored before trained model configuration"
+            ));
+            return;
+        }
+        TrainedModelDefinitionDoc trainedModelDefinitionDoc = trainedModelDefinitionChunk.createTrainedModelDoc(this.currentModelId.get());
+
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<Void> storeListener = ActionListener.wrap(
+            r -> {
+                LOGGER.debug(() -> new ParameterizedMessage(
+                    "[{}] stored trained model definition chunk [{}] [{}]",
+                    analytics.getId(),
+                    trainedModelDefinitionDoc.getModelId(),
+                    trainedModelDefinitionDoc.getDocNum()));
+
+                if (trainedModelDefinitionChunk.isEos()) {
+                    readyToStoreNewModel = true;
+                    LOGGER.info(
+                        "[{}] finished stored trained model definition chunks with id [{}]",
+                        analytics.getId(),
+                        this.currentModelId.get());
+                    auditor.info(analytics.getId(), "Stored trained model with id [" + this.currentModelId.get() + "]");
+                    CountDownLatch refreshLatch = new CountDownLatch(1);
+                    provider.refreshInferenceIndex(
+                        new LatchedActionListener<>(ActionListener.wrap(
+                            refreshResponse -> LOGGER.debug(() -> new ParameterizedMessage(
+                                "[{}] refreshed inference index after model store",
+                                analytics.getId()
+                            )),
+                            e -> LOGGER.warn("[{}] failed to refresh inference index after model store", analytics.getId())),
+                            refreshLatch));
+                    try {
+                        if (refreshLatch.await(30, TimeUnit.SECONDS) == false) {
+                            LOGGER.error("[{}] Timed out (30s) waiting for index refresh", analytics.getId());
+                        }
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
+                }
+            },
+            e -> failureHandler.accept(ExceptionsHelper.serverError("error storing trained model definition chunk [{}] with id [{}]", e,
+                trainedModelDefinitionDoc.getModelId(), trainedModelDefinitionDoc.getDocNum()))
+        );
+        provider.storeTrainedModelDefinitionDoc(trainedModelDefinitionDoc, new LatchedActionListener<>(storeListener, latch));
```

Comment:
Add a boolean parameter here to refresh the index when the last doc is indexed that way you don't have to call  `provider.refreshInferenceIndex()` in the listener

Reply:
I am refactoring this so that the latched listener is the refresh listener and the provider listener always calls the refresh listener (with `null` when we don't refresh).